localstack/services/dynamodb/dynamodb_starter.py

```diff
@@ -1,81 +1,57 @@
-    get_free_tcp_port,
-    mkdir,
-    wait_for_port_closed,
-    wait_for_port_open,
-)
 
-LOGGER = logging.getLogger(__name__)
+LOG = logging.getLogger(__name__)
 
-# backend service port (updated on startup)
-PORT_DYNAMODB_BACKEND = None
-
-# todo: will be replaced with plugin mechanism
-PROCESS_THREAD = None
+# server singleton
+_server: Optional[DynamodbServer] = None
 
 
 def check_dynamodb(expect_shutdown=False, print_error=False):
     out = None
+
+    if not expect_shutdown:
+        assert _server
+
     try:
-        # wait for backend port to be opened
-        wait_for_port_open(PORT_DYNAMODB_BACKEND, http_path="/", expect_success=False, sleep_time=1)
-        # check DynamoDB
-        endpoint_url = f"http://127.0.0.1:{PORT_DYNAMODB_BACKEND}"
-        out = aws_stack.connect_to_service("dynamodb", endpoint_url=endpoint_url).list_tables()
+        _server.wait_is_up()
+        out = aws_stack.connect_to_service("dynamodb", endpoint_url=_server.url).list_tables()
     except Exception:
         if print_error:
-            LOGGER.exception("DynamoDB health check failed")
+            LOG.exception("DynamoDB health check failed")
     if expect_shutdown:
         assert out is None
     else:
         assert isinstance(out["TableNames"], list)
 
 
-def start_dynamodb(port=None, asynchronous=False, update_listener=None):
-    global PROCESS_THREAD, PORT_DYNAMODB_BACKEND
-    PORT_DYNAMODB_BACKEND = get_free_tcp_port()
-    port = port or config.PORT_DYNAMODB
-    install.install_dynamodb_local()
-    ddb_data_dir_param = "-inMemory"
-    if config.DATA_DIR:
-        ddb_data_dir = "%s/dynamodb" % config.DATA_DIR
-        mkdir(ddb_data_dir)
-        # as the service command cds into a different directory, the absolute
-        # path of the DATA_DIR is needed as the -dbPath
-        absolute_path = os.path.abspath(ddb_data_dir)
-        ddb_data_dir_param = "-dbPath %s" % absolute_path
-    cmd = (
-        "cd %s/infra/dynamodb/; java -Djava.library.path=./DynamoDBLocal_lib "
-        + "-Xmx%s -jar DynamoDBLocal.jar -port %s %s"
-    ) % (
-        MODULE_MAIN_PATH,
-        config.DYNAMODB_HEAP_SIZE,
-        PORT_DYNAMODB_BACKEND,
-        ddb_data_dir_param,
-    )
+def start_dynamodb(port=None, asynchronous=True, update_listener=None):
+    global _server
+    if not _server:
+        _server = create_dynamodb_server()
+
+    _server.start()
+
     log_startup_message("DynamoDB")
     start_proxy_for_service(
         "dynamodb",
         port,
-        backend_port=PORT_DYNAMODB_BACKEND,
+        backend_port=_server.port,
         update_listener=update_listener,
     )
-    # todo: extract reference from do_run (should return pid)
-    PROCESS_THREAD = do_run(cmd, asynchronous, auto_restart=True)
-    return PROCESS_THREAD
+    return _server
 
 
 def restart_dynamodb():
-    LOGGER.debug("Restarting DynamoDB process ...")
-    PROCESS_THREAD.stop()
-    wait_for_port_closed(PORT_DYNAMODB_BACKEND)
+    global _server
```

Comment:
nit: `global` keyword probably not required, if we're not assigning a new value to `_server`?

Reply:
yep good catch :-)